File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        HashSet<ListNode> seenNodes = new HashSet<>();
        ListNode dummy = new ListNode(head.val, head.next);
        while (dummy != null) {
            if (seenNodes.contains(dummy)) {
                return true;
            }
            seenNodes.add(dummy);
            dummy = dummy.next;
        }
        return false;
        
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        length = 0
        curr = head

        while curr:
            length += 1
            curr = curr.next

        pos_from_start = length - n

        prev = dummy
        for i in range(pos_from_start):
            prev = prev.next

        prev.next = prev.next.next

        return dummy.next